cirq/sim/simulator.py

```diff
@@ -359,7 +366,11 @@ def simulate_sweep(
         raise NotImplementedError()
 
 
-class SimulatesIntermediateState(SimulatesFinalState, metaclass=abc.ABCMeta):
+class SimulatesIntermediateStateBase(
```

Comment:
Why is a `*Base` type needed here, but not for the other newly-generic types?

Reply:
Because `SimulatesIntermediateState` defines a `_create_simulator_trial_result` that returns a `SimulationTrialResult`. But some child classes return some child result type.

Reply:
Ah - it has type-specific implementation, but the others don't. That makes sense.

Should the other types then be `{Type}Base` for consistency? Or perhaps it's cleaner to leave the "Base" versions as-is, and rename the new `SimulatesIntermediateState` to `SimulatesIntermediateStateImpl`?

Reply:
Probably the latter. And we may even be able to delete the `Impl` afterward because I think it's only used in a couple tests. What do you think?

Reply:
That sounds good to me, although to preserve test coverage I'd suggest moving the `Impl` into the tests instead of deleting it entirely.

Reply:
done